### DISCUSSION TITLE
Best strategies for multi-GPU and multi-CPU simulations?

### DISCUSSION THREAD

**Discussion by ali-ramadhan** [CliMA/Oceananigans.jl#3885] [High performance computing]
Been digging into the distributed and multi-region modules (really awesome work!) and I'm curious about the best strategies for running large simulations on multiple CPUs and GPUs.

I know @simone-silvestri has run simulations on hundreds to thousands (!!) of GPUs. I'm thinking more of the case of using 2-16 (_maybe_ up to 64) GPUs to run hydrostatic and non-hydrostatic simulations on rectilinear and lat-lon grids.

Here's what I think I've figured out so far, but would love to hear any thoughts or experience from others.

* For multi-GPU on a single node (usually up to 4-8 GPUs), it seems like you should use multi-regions to avoid touching MPI?

* For multi-GPU on multiple nodes, is it possible to use multi-region on each node then use CUDA-aware MPI between nodes (one rank per node)?

* For multiple cores on a single CPU on one node, I know we can either use multi-threading via KernelAbstractions.jl or use MPI. If I remember correctly MPI scales better here?

* For multi-CPU on multiple nodes, has anyone tried doing this? Going full MPI (one rank per CPU core) seems like you'll end up with too many ranks and too much communication. So maybe the winning strategy is multi-threading with KernelAbstractions.jl on each node and MPI between nodes (one rank per node).

I don't think multi-CPU is really of much use or interest as you probably need multiple beefy CPUs to rival just one GPU (and deal with MPI), but I'm mentioning it out of personal curiosity I suppose.

**Comment by glwagner** [comment]
The cubed sphere is the only application for MultiRegion.

 > If I remember correctly MPI scales better here?

This seems very unlikely, but could be checked. I suspect that multi-threading is the way to go.

> For multi-CPU on multiple nodes, has anyone tried doing this?

Not that I know of.

> So maybe the winning strategy is multi-threading with KernelAbstractions.jl on each node and MPI between nodes (one rank per node).

Agreed.

> I don't think multi-CPU is really of much use or interest as you probably need multiple beefy CPUs to rival just one GPU (and deal with MPI)

KernelAbstractions is also non-optimal for CPUs.

**Comment by XinghaoJ** [comment]
My simulation is currently constrained by GPU memory limitations, so I'm considering utilizing multiple GPUs. I have a few questions regarding multi-GPU simulations:

1. If multi-regions can be employed, is it possible to conduct our calculations within a single node (but with 4-8 GPUs) to avoid using MPI?
2. Is there a method to implement multi-regions in RectilinearGrid grids? It appears that the command MultiRegionGrid(global_grid; partition = XPartition(2), devices = nothing, validate = true) might be applicable.

**Reply by glwagner** [reply]
As preface: to use multi-GPUs we recommend utilizing the `Distributed` architecture with MPI. 

1. No, multi-GPU simulations without MPI are not supported, unfortunately. We have found that it is too much work to support more than one paradigm for multi-device simulations.

2. MultiRegion simulations on anything but the cubed sphere have not been tested or developed. However, such development work is welcome if you would like to undertake it.

**Reply by XinghaoJ** [reply]
Thanks for your quick response! I have a fellowed question, after I distribute the mesh and use halo to convey information between partitioned parts of fields, do we need to use "callback" command to mannually update the Fields at halo at every timestep? can this model itselt automatically convey the Fields information at halo during the calculation of every timestep?  Specifically, when I bulid a model, is that right to directly code with 

```julia
child_architecture = GPU()
architecture = Distributed(child_architecture)
grid = RectilinearGrid(architecture,
                       size = (48, 48, 16),
                       x = (0, 64),
                       y = (0, 64),
                       z = (0, 16),
                       topology = (Periodic, Periodic, Bounded))
boundary_conditions = ...
NonhydrostaticModel(; grid, boundary_conditions=boundary_conditions ....)
```

or I need to build  NonhydrostaticModel for every partitioned part of mesh?

**Reply by glwagner** [reply]
No --- filling halos is performed automatically (this is also true for the serial case, which requires halo filling for most boundary condition types).

The main change that is required in your script is to change the architecture. (If that's not the case for some reason, please open an issue and we can address it.)

**Reply by XinghaoJ** [reply]
Thanks a lot, I'm currently trying about this.
